File: apps/web/lib/api/widget-keys.ts
import "server-only";

import { cache } from "react";

import { apiFetch } from "./client";

/**
 * A console-managed widget embed (spec D1/D4). Admin-only, reached under
 * `/widget-keys` -- not to be confused with `lib/api/widget.ts`, the
 * anonymous door the embedded widget itself calls by key.
 *
 * Every schema on the API side inherits `CamelModel`, so the wire is already
 * camelCase and needs no mapping layer here.
 */
export interface WidgetKey {
  id: string;
  name: string;
  /**
   * Returned in full on every read, unlike `ApiKey`'s one-time token. This
   * value is public by construction -- it is meant to sit in the customer's
   * page source, so there is nothing to protect by hiding it after creation.
   */
  key: string;
  allowedOrigins: string[];
  /** Launcher colour, position, greeting and the like. Deliberately not
   * rendered anywhere in this slice -- see the widget settings page brief. */
  settings: Record<string, unknown>;
  active: boolean;
  lastSeenAt: string | null;
  createdAt: string;
}

export interface WidgetKeyInput {
  name: string;
  allowedOrigins?: string[];
}

export const getWidgetKeys = cache(async (): Promise<WidgetKey[]> => {
  return apiFetch<WidgetKey[]>("/widget-keys");
});

export async function createWidgetKey(input: WidgetKeyInput): Promise<WidgetKey> {
  return apiFetch<WidgetKey>("/widget-keys", {
    method: "POST",
    body: JSON.stringify(input),
  });
}

export async function updateWidgetKey(
  id: string,
  patch: Partial<WidgetKeyInput> & { active?: boolean },
): Promise<WidgetKey> {
  return apiFetch<WidgetKey>(`/widget-keys/${id}`, {
    method: "PATCH",
    body: JSON.stringify(patch),
  });
}

export async function deleteWidgetKey(id: string): Promise<void> {
  await apiFetch(`/widget-keys/${id}`, { method: "DELETE" });
}

File: apps/api/src/relaydesk/schemas/widget_keys.py
import uuid
from datetime import datetime

from pydantic import Field

from relaydesk.schemas.base import CamelModel


class WidgetKeyCreate(CamelModel):
    name: str = Field(min_length=1, max_length=120)
    allowed_origins: list[str] = Field(default_factory=list)
    settings: dict = Field(default_factory=dict)


class WidgetKeyUpdate(CamelModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    allowed_origins: list[str] | None = None
    settings: dict | None = None
    active: bool | None = None


class WidgetKeyOut(CamelModel):
    id: uuid.UUID
    name: str
    # Returned in full on every read, unlike ApiKey's one-time secret: this
    # value is published in page source and the admin needs it to re-paste.
    key: str
    allowed_origins: list[str]
    settings: dict
    active: bool
    last_seen_at: datetime | None
    created_at: datetime
